### ISSUE TITLE
the derivative for 'cudnn_batch_norm_backward save_mean' is not implemented

### ISSUE THREAD

**Issue by elbaro** [pytorch/pytorch#4422]
```
loss_gp = ..
loss_gp.backward()  # error
```

Having this error when BatchNorm + wgan_gp (second order) is used.
It's installed from master fec3d4.

**Comment by ezyang** [comment]
This is almost definitely my fault. I'll look into it on Tuesday.

**Comment by ezyang** [comment]
@elbaro Would it be possible to get more detailed reproduction instructions?

**Comment by ezyang** [comment]
Never mind, it's pretty easy to repro:

```
    def test_batchnorm_double_backward(self):
        bn = nn.BatchNorm1d(5).cuda()
        v = Variable(torch.randn(5, 5), requires_grad=True).cuda()
        output = bn(v)
        grad, = torch.autograd.grad(outputs=output.sum(), inputs=v, create_graph=True, retain_graph=True, only_inputs=True)
        grad.sum().backward()
```

**Comment by ezyang** [comment]
Root causes:

- I did a hack to work around lack of ability to return non-requires grad outputs from ATen functions, but apparently this hack doesn't work. We need a way of returning variables from an ATen function such that they unconditionally don't require grad. @colesbury suggested that we add a new special form to `derivatives.yaml` to trigger this but I am actually not too sure we can actually make this work. In any case, @colesbury has some PRs in progress on this code so I won't be able to fix this today.
- Why didn't I catch the hack failure? We are apparently not running any NewModuleTests double-backwards with CUDA. Oops. I'm going to try to fix this today.

To work around the problem:
- Turn off cuDNN with `torch.backends.cudnn.enabled = False`, OR
- Revert to PyTorch d7e6ede78465ceddf574840eb844f08a9e870557 or earlier
